src/integer_mod_q/mat_zq/solve.rs

```diff
@@ -58,44 +62,314 @@ impl MatZq {
+                column_nr,
+                used_rows.clone(),
+            ) {
+                let row = inv * matrix_identity_base_gauss.get_row(row_nr).ok()?;
+                matrix_identity_base_gauss.set_row(row_nr, &row, 0).ok()?;
+
+                // Set all other entries in that column to `0` (gaussian elimination).
+                for row_nr in
+                    (0..matrix_identity_base_gauss.get_num_rows()).filter(|x| *x != row_nr)
+                {
+                    let old_row = matrix_identity_base_gauss.get_row(row_nr).ok()?;
+                    let entry: Z = old_row.get_entry(0, column_nr).ok()?;
+                    let new_row = &old_row - entry * &row;
+                    matrix_identity_base_gauss
+                        .set_row(row_nr, &new_row, 0)
+                        .ok()?;
+                }
+
+                if row_count != row_nr {
+                    matrix_identity_base_gauss
+                        .swap_rows(row_count, row_nr)
+                        .unwrap();
+
+                    permutation.swap(row_count.try_into().unwrap(), row_nr.try_into().unwrap());
+                }
+
+                // Save the position of `1` for that row.
+                indices.push((row_count, column_nr));
+                row_count += 1;
+            // Search for a not invertible entry to divide the modulus.
+            } else if let Some(row_nr) =
+                find_not_invertible_entry_column(&matrix_identity_base_gauss, column_nr, used_rows)
+            {
+                // Factorize the modulus with the found entry, compute solutions
+                // for the system under the split modulus and use the
+                // Chinese Remainder Theorem to compute a solution based on these
+                // sub-solutions.
+                let entry: Z = matrix_identity_base_gauss
+                    .get_entry(row_nr, column_nr)
+                    .ok()
+                    .unwrap();
+                self.factorization_and_crt(y, entry);
+            }
         }
 
-        // check if we have computed a solution and return `None`
-        if &(self * &out) != y {
+        // Return `None` if the matrix has no full rank.
+        if indices.is_empty()
+            || indices[indices.len() - 1].0 + 1 < matrix_identity_base_gauss.get_num_rows()
+        {
             return None;
         }
 
+        // Pick the first columns out of the original matrix that form an invertible matrix.
+        // Those columns exist, since the matrix was checked to be full rank.
+        let mut square_matrix = MatZq::new(
+            matrix_identity_base_gauss.get_num_rows(),
+            matrix_identity_base_gauss.get_num_rows(),
+            self.get_mod(),
+        );
+        for (current_column, (_row_nr, column_nr)) in indices.iter().enumerate() {
+            square_matrix
+                .set_column(current_column, self, *column_nr)
+                .unwrap();
+        }
```

Comment:
the name square matrix is misleading

Reply:
I changed it to invertible_matrix, this should clarify it